src/Core/Services/Implementations/PolicyService.cs

```diff
@@ -53,25 +54,52 @@ public class PolicyService : IPolicyService
                         {
-                            if (orgUser.UserId != savingUserId && !await userService.TwoFactorIsEnabledAsync(orgUser))
+                            switch (currentPolicy.Type)
                             {
-                                if (organization == null)
-                                {
-                                    organization = await _organizationRepository.GetByIdAsync(policy.OrganizationId);
-                                }
-                                await organizationService.DeleteUserAsync(policy.OrganizationId, orgUser.Id,
-                                    savingUserId);
-                                await _mailService.SendOrganizationUserRemovedForPolicyTwoStepEmailAsync(
-                                    organization.Name, orgUser.Email);
+                                case Enums.PolicyType.TwoFactorAuthentication:
+                                    if (!await userService.TwoFactorIsEnabledAsync(orgUser))
+                                    {
+                                        if (organization == null)
+                                        {
+                                            organization = await _organizationRepository.GetByIdAsync(policy.OrganizationId);
+                                        }
+                                        await organizationService.DeleteUserAsync(policy.OrganizationId, orgUser.Id,
+                                            savingUserId);
+                                        await _mailService.SendOrganizationUserRemovedForPolicyTwoStepEmailAsync(
+                                            organization.Name, orgUser.Email);
+                                    }
+                                    break;
+                                case Enums.PolicyType.OnlyOrg:
+                                    if (orgUser.UserId.HasValue)
+                                    {
+                                        var userOrgs = await _organizationUserRepository.GetManyByUserAsync(orgUser.UserId.Value);
+                                        if (userOrgs.Count > 1)
+                                        {
+                                            if (organization == null)
+                                            {
+                                                organization = await _organizationRepository.GetByIdAsync(policy.OrganizationId);
+                                            }
+                                            await organizationService.DeleteUserAsync(policy.OrganizationId, orgUser.Id,
+                                                savingUserId);
+                                            await _mailService.SendOrganizationUserRemovedForPolicyOnlyOrgEmailAsync(
+                                                organization.Name, orgUser.Email);
+                                        }
```

Comment:
Seems like this operation could get expensive in a loop of possibly thousands of users. Any way to optimize further?

Reply:
For sure. Setting up some kind of bulk GET stored procedure that accepts a list of orgUser Ids and gets all of those at once would cut trips to the DB for that operation down to one, but there doesn't seem to be a precedent for a "get by many" type of procedure so far, so let me know if that's the direction you'd like to go before I make the swap, @kspearrin .

Reply:
@addisonbeck , there's a UDT in the database, `GuidIdArray` which is basically a `TABLE` type with a single column, `[Id] UNIQUEIDENTIFIER NOT NULL` that you an use for creating a bulk stored procedure that takes in an `IEnumerable<Guid>` and returns the data in a set operation (`JOIN @tableVariable AS Ids ON -- etc...`).

See `[dbo].[Cipher_UpdateCollections]` for a "kind of" example where this is used (at least as a sproc parameter and Dapper mapper, the use-case it a bit difference once you have it in the sproc).

Reply:
Done!